test/lower/expr-test-generator.cc

```diff
@@ -0,0 +1,692 @@
+
+// Sorted by increasing precedence (not stricly increasing). See F2018
+// table 10.1.
+static Operation operations[]{
+    // TODO should be OK constexpr but g++ 8.2 not happy
+    // defined-binary-op cannot be described here.
+    {"eqv", ".EQV.", TypePattern::OperandLogical},
+    {"neqv", ".NEQV.", TypePattern::OperandLogical},
+    {"or", ".OR.", TypePattern::OperandLogical},
+    {"and", ".AND.", TypePattern::OperandLogical},
+    {"not", ".NOT.", TypePattern::OperandLogical, Constraint::None, true},
+    {"eq", ".EQ.", TypePattern::OperandComparable},
+    {"ne", ".NE.", TypePattern::OperandComparable},
+    {"lt", ".LT.", TypePattern::OperandOrdered},
+    {"le", ".LE.", TypePattern::OperandOrdered},
+    {"gt", ".GT.", TypePattern::OperandOrdered},
+    {"ge", ".GE.", TypePattern::OperandOrdered},
+    // TODO: Concat
+    {"add", "+", TypePattern::OperandNumeric},
+    {"sub", "-", TypePattern::OperandNumeric},
+    {"minus", "-", TypePattern::OperandNumeric, Constraint::None, true},
+    {"plus", "+", TypePattern::OperandNumeric, Constraint::None, true},
+    {"mult", "*", TypePattern::OperandNumeric},
+    {"div", "/", TypePattern::OperandNumeric, Constraint::NonZeroRHS},
+    {"power", "**", TypePattern::OperandNumeric,
+        Constraint::NoNegativeLHSWhenRealRHS},  // TODO: check power pattern
+    // defined-unary-op cannot be described here.
+};
+
+static bool IsLogical(const Type t) { return t.cat == TypeCategory::Logical; }
+static bool IsCharacter(const Type t) {
+  return t.cat == TypeCategory::Character;
+}
+static bool IsComplex(const Type t) { return t.cat == TypeCategory::Complex; }
+static bool IsReal(const Type t) { return t.cat == TypeCategory::Real; }
+
+static bool IsOrderedNumeric(const Type t) {
+  return t.cat == TypeCategory::Real || t.cat == TypeCategory::Integer;
+}
+
+static bool IsNumeric(const Type t) {
+  return t.cat == TypeCategory::Real || t.cat == TypeCategory::Integer ||
+      t.cat == TypeCategory::Complex;
+}
+
+std::optional<Type> Operation::GetResultTypeForArgument(Type t) const {
+  switch (pattern) {
+  case TypePattern::OperandNumeric:
+    if (IsNumeric(t)) {
+      return t;
+    }
+    break;
+  case TypePattern::OperandLogical:
+    if (IsLogical(t)) {
+      return t;
+    }
+    break;
+  case TypePattern::OperandComparable:
+  case TypePattern::OperandOrdered:
+    assert(false && "No unary comparisons");
+    break;
+  }
+  return std::nullopt;
+}
+
+static std::optional<Type> SelectBiggestKindForResult(
+    TypeCategory resultCat, Type lhs, Type rhs) {
+  if (!lhs.kind && !rhs.kind) {
+    return Type{resultCat};
+  } else if (lhs.kind && rhs.kind) {
+    return Type{resultCat, std::max(*lhs.kind, *rhs.kind)};
+  } else {
+    // Cannot know without more info about compiler what default compares to.
+    return std::nullopt;
+  }
+}
+
+std::optional<Type> Operation::GetResultTypeForArguments(
+    Type lhs, Type rhs) const {
+  // See Fortran 2018 table 10.2 and section 10.1.9.3
```

Comment:
I believe in the future would be nice to write the reason and reference the standard for more info.
Instead of only adding the reference to the standard